internal/writeaheadlog/wal.go

```diff
@@ -0,0 +1,295 @@
+package writeaheadlog
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"go.uber.org/multierr"
+)
+
+var log = logging.Logger("f3/wal")
+
+type Entry interface {
+	WALEpoch() uint64
+	cbg.CBORMarshaler
+	cbg.CBORUnmarshaler
+}
+
+type WriteAheadLog[T any, PT interface {
+	*T
+	Entry
+}] struct {
+	// pointer type trickery above
+	lk sync.Mutex
+
+	path string
+
+	logFiles []logContent[T, PT]
+	active   struct {
+		file       *os.File
+		cborWriter *cbg.CborWriter
+		logContent[T, PT]
+	}
+}
+
+type logContent[T any, PT interface {
+	*T
+	Entry
+}] struct {
+	logname  string
+	content  []T
+	maxEpoch uint64
+	complete bool
+}
+
+func Open[T any, PT interface {
+	*T
+	Entry
+}](directory string) (*WriteAheadLog[T, PT], error) {
+	wal := WriteAheadLog[T, PT]{
+		path: directory,
+	}
+
+	err := wal.readInAll()
+	if err != nil {
+		return nil, fmt.Errorf("reading the WAL: %v", err)
+	}
+
+	return &wal, nil
+}
+
+func (wal *WriteAheadLog[T, PT]) All() []T {
```

Comment:
`All` is convention. But, now that we have iterators, it would be kind of nice to use them here: https://medium.com/eureka-engineering/a-look-at-iterators-in-go-f8e86062937c (especially if we stop storing these in memory).